### PR TITLE
Implement loading modules from piped files

### PR DESCRIPTION
**Pull request by cr0sh**
This is not discussed in the issue tracker, but on this [Zulip thread](https://bytecodealliance.zulipchat.com/#narrow/stream/217126-wasmtime/topic/Reading.20module.20from.20stdin/near/302484255).

Previously load_module opened a file on `path` to check if the module is precompiled, then closes before constructing the Module from it. This is fine on normal files, but on piped files(like /dev/stdin), is problematic as the header read is not visible on the next open.

So, this set of commits inserts a new procedure, before pre-opening to read the header, to detect if the file pointed by the `path` is a pipe([`is_fifo`](https://doc.rust-lang.org/stable/std/os/unix/fs/trait.FileTypeExt.html#tymethod.is_fifo) of FileTypeExt), then directly loads the module from the path without checking if the modules is precompiled.

This change is **unix-only**, and intentionally does not support loading precompiled modules from pipes, because it should not be used that way. An additional warning message is printed in such situations to prevent users from being confused.

The Zulip thread contains an additional concern by me, about implicitly allowing accsesing `/dev/tty` on `-` inputs, which is not considered here.

*Edited to describe the current procedure to detect piped files.

### CODE REVIEW

**Comment by cr0sh** [comment]
It seems like I cannot assign a reviewer for the PR(the template indicates me to do so), so not assigning.

File: src/commands/run.rs [hunk 1]
```diff
@@ -423,22 +426,32 @@ impl RunCommand {
...
+        // If it fails(it wouldn't be a regular file), allocate a fixed mmap buffer
+        // to store the whole bytes in a memory.
+        let mmap_vec = MmapVec::from_file(path).or_else(|_| {
+            // Mmap-ing this much would not eat system memory thanks for paging
+            let mut buf = Vec::new();
+            let mut f = File::open(path)?;
+            f.read_to_end(&mut buf)?;
+            let mmap_vec = MmapVec::from_slice(&buf)?;
+            Ok::<_, anyhow::Error>(mmap_vec)
+        })?;
```
Comment (bjorn3) [inline src/commands/run.rs:440]: This introduces new copies of the data, slowing down startup. Previously `Module::from_file` would mmap the file once and then use this mmap'ed version directly. Now it will mmap it once (or read to a vec and then copy it to a new memory region in `MmapVec::from_slice`) and then copy it to a new memory region again in `Module::new` (through `MmapVec::from_slice`).

Comment (cr0sh) [reply src/commands/run.rs:440]: Definitely right. To be honest, I thought directly writing Mmap/MmapVec with looped `f.read()` is more efficient but feeled like a premature optimization, so chose this implementation for brevity. (The main reason for this is the File::read_to_end is not compatible with MmapVec.)

If startup performance matters for piped inputs to, I'll fix this into like
```rust
loop {
    let buf = [0u8; 4<<10];
    match f.read(&mut buf) {
        Ok(0) => break,
        Ok(n) => /* copy `n` bytes to Mmap and increment the offset */ todo!(),
        Err(e) => return Err(e),
    }
}
```

But again I'm not sure this is worth it.

**Comment by cr0sh** [comment]
Piping tests now run on unix platforms only(as the `-` substitiution applies only on unix too). Double buffering issue is resolved.
What do you think about implicitly allowing access to /dev/tty on `wasmtime -` as a special case?

File: src/commands/run.rs (hunk 1 above)
Comment (bjorn3) [reply src/commands/run.rs:454]: This will still have an extra copy in case the mmap succeeded.

Comment (cr0sh) [reply src/commands/run.rs:454]: OK I didn't understand what you were saying. 😅 Now I see.
I'm afraid that this is not able to be trivially fixed, as there's no public API on `Module` to pass MmapVec directly into a constructor?

`from_binary` seems to be the 'generic' catch-all for many Module constructors but I don't have an idea what it does...

Comment (cr0sh) [reply src/commands/run.rs:454]: Nevermind, you said about the case when the mmap suceeded, so trying to mmap and if it succeeds then discarding the mmap and use Module::from_path is OK. Is that you're thinking?

Comment (bjorn3) [reply src/commands/run.rs:454]: That may work.

File: src/commands/run.rs [hunk 2]
```diff
@@ -423,22 +426,46 @@ impl RunCommand {
...
-            if &magic == b"\x7fELF" {
-                if self.allow_precompiled {
-                    return unsafe { Module::deserialize_file(engine, path) };
+
+        // Try to mmap the module file.
+        // If it fails(it wouldn't be a regular file), allocate a fixed mmap buffer
+        // to store the whole bytes in a memory.
+        let mmap_vec = MmapVec::from_file(path).or_else(|_| {
+            // Mmap-ing this much would not eat system memory thanks for paging
+            const MAX_READ_FROM_STREAM: usize = 1 << 30; // TODO: make this configurable with environment variables?
```
Comment (sunfishcode) [inline src/commands/run.rs:435]: Arbitrary limitations and environment variables are something we'd like to avoid if possible. The users of `mmap_vec` don't care if they have an `MMapVec` or a plain `Vec<u8>`, so in the case where the mmap fails, would it work to just call `std::fs::read` to read the file? That would require reorganizing this code to avoid using `MmapVec` as the type to unify the two options. Something like this:

```rust
    let mmap_vec;
    let read;
    let bytes: &[u8] = if let Ok(mapped) = MmapVec::from_file(Path::new(path)) {
        mmap_vec = mapped;
        mmap_vec.as_ref()
    } else {
        read = std::fs::read("hello.txt")?;
        read.as_ref()
    };
```

Comment (cr0sh) [reply src/commands/run.rs:435]: Yeah I misinterpreted @bjorn3 's first review so I'll roll back into previous implementation which just uses `Read::read_to_end` and `Vec<u8>`, and prevent redundant allocation when a module is from a regular file. This *may* be sub-optimal for irregular file case but I think it's OK because no one wants to do `hyperfine 'cat foo.wasm | wasmtime -'`?

File: Cargo.toml [hunk 3]
```diff
@@ -30,6 +30,7 @@ wasmtime-wast = { workspace = true }
 wasmtime-wasi = { workspace = true }
 wasmtime-wasi-crypto = { workspace = true, optional = true }
 wasmtime-wasi-nn = { workspace = true, optional = true }
+wasmtime-runtime = { workspace = true }
```
[the remark this answers is not in the document]
Comment (cr0sh) [reply Cargo.toml:33]: This was only needed for MmapVec so I'll remove on the next commit. (see [below](https://github.com/bytecodealliance/wasmtime/pull/5035#discussion_r992230460).)

[the remark this answers is not in the document]
Comment (cr0sh) [reply Cargo.toml:33]: After re-reading my previous commits and review comments here, I realized this reply is not possible at all. MmapVec cannot be avoided because `load_module` should do both "header check" and "copyless mmap" in one path. I think your suggestion is truly valid but please understand that I don't know much about this project (yet) to implement it rapidly so I would come back with results after some time with some investigation and tryouts.

[the remark this answers is not in the document]
Comment (sunfishcode) [reply Cargo.toml:33]: Sounds good. And thanks for working on this! I agree that running modules from stdin is an interesting idea.

**Comment by cr0sh** [comment]
I simplified the strategy to detect piped files(directly invokes `is_fifo` on Unix) and fixed tests incorrectly set up.

File: src/commands/run.rs [hunk 4]
```diff
@@ -428,6 +428,7 @@ impl RunCommand {
 
         // Currently loading modules from pipes are an unix-only feature
         if cfg!(unix) {
+            #[cfg(unix)]
```
Comment (bjorn3) [inline src/commands/run.rs:431]:
```suggestion
        #[cfg(unix)]
        {
```

**Comment by jameysharp** [comment]
Here's the strategy I'd suggest:

Add a new method to `wasmtime::Module` that accepts both precompiled binaries and wasm/wat files, and uses `mmap`. I'd call it `from_trusted_file` but I'm open to suggestions. It should probably be `unsafe`, like `Module::deserialize_file`, for the same reasons.

In the CLI, if `allow_precompiled` is true, then call `Module::from_trusted_file`; otherwise, call `Module::from_file` just like it does today. Don't inspect the bytes of the file at all in the CLI, but put that logic in the `wasmtime` crate instead. I think that addresses @sunfishcode's comment on keeping the CLI as a simple demonstration of how to use the `wasmtime` public API, and improves on the current state where there's this magic ELF-header check in the CLI.

In your motivating use case (`curl | wasmtime`), users shouldn't be passing `--allow-precompiled`, so the CLI would call `Module::from_file`, resulting in a single `fs::read` that consumes the piped input exactly once, which solves the original bug you encountered.

This new method should read the file using `MmapVec::from_file`. Only the OS knows exactly which file types it's capable of mapping into memory, so this shouldn't check whether the file is a FIFO or anything along those lines; just let the syscall fail if someone uses `--allow-precompiled` with a pipe.

Once you have an `MmapVec`, you can check whether it begins with `b"\x7fELF"`. If so, pass it to `SerializedModule::from_mmap`; otherwise, pass it to `Module::new`. Neither function copies the bytes it's given, addressing @bjorn3's concern.

I think the only downside to this plan is that the error message from the CLI isn't as friendly if someone tries to run a precompiled artifact without passing `--allow-precompiled`. We could make `Module::build_artifacts` return a more helpful error if the bytes it receives start with an ELF header, instead of just reporting that parsing failed.

One advantage to this approach is that `wasmtime --allow-precompiled` becomes potentially faster for .wat and .wasm files, because it'll use mmap for those cases too.

**Comment by cr0sh** [comment]
@jameysharp Thanks for your detailed guidance! I didn't think fixing this would not be hard this way and admittedly I couldn't do some sort of refactoring the project(and even adding a new public API for it) never saw before, tempting myself to bring somewhat hackish way. I'll try to reimplement this way.

> I'd call it from_trusted_file but I'm open to suggestions.

This looks good.

> I think the only downside to this plan is that the error message from the CLI isn't as friendly if someone tries to run a precompiled artifact without passing --allow-precompiled. We could make Module::build_artifacts return a more helpful error if the bytes it receives start with an ELF header, instead of just reporting that parsing failed.

I'm not sure that I can bring an excellent solution for this; just will do my best.

To recap I leave a tabular summary for your suggestion:

|     | no `--allow-precompiled` <br /> (`Module::from_file`)  | `--allow-precompiled` <br /> (`Module::from_trusted_file`) |
|--|--|--| 
| precompiled | regular: opens OK and fails on some inner procedure but should be rejected via an additional check? TODO: better error handling <br /> fifo: same | regular: opens OK w/ mmap <br /> fifo: mmap fails (intentionally left broken) | 
| `.wasm`/`.wat` | regular: opens OK <br /> fifo: **opens OK** (major change on this PR) | regular: opens OK w/mmap <br /> fifo: same |

**Comment by cr0sh** [comment]
Rebased and implemented the @jameysharp 's guidance. (I'll move to a new pull request as the topic has changed, so please leave this PR not reviewed)
